netfilter: add kmod-nft-dup-inet

kwiateusz opened the pull request:
Several people, as I found on google, had the same problem as me. They wanted to duplicate packets, for example between vlans.

The only possibility was to compile the image on my own, this package will solve this problem.

john-tho commented on package/kernel/linux/modules/netfilter.mk:
```diff
@@ -1147,6 +1147,30 @@ endef
+define KernelPackage/nft-dup
+  SUBMENU:=$(NF_MENU)
+  TITLE:=Netfilter nf_tables dup support
+  DEPENDS:=+kmod-nft-core
+  KCONFIG:= \
+  CONFIG_NF_DUP_NETDEV \
+  CONFIG_NF_DUP_IPV4 \
+  CONFIG_NF_DUP_IPV6 \
+  CONFIG_NFT_DUP_NETDEV \
+  CONFIG_NFT_DUP_IPV4 \
+  CONFIG_NFT_DUP_IPV6
+  FILES:= \
+  $(LINUX_DIR)/net/netfilter/nf_dup_netdev.ko \
+  $(LINUX_DIR)/net/ipv4/netfilter/nf_dup_ipv4.ko \
+  $(LINUX_DIR)/net/ipv6/netfilter/nf_dup_ipv6.ko \
+  $(LINUX_DIR)/net/netfilter/nft_dup_netdev.ko \
+  $(LINUX_DIR)/net/ipv4/netfilter/nft_dup_ipv4.ko \
+  $(LINUX_DIR)/net/ipv6/netfilter/nft_dup_ipv6.ko
+  AUTOLOAD:=$(call AutoProbe,nf_dup_netdev nf_dup_ipv4 nf_dup_ipv6 nft_dup_netdev nft_dup_ipv4 nft_dup_ipv6)
```
There is already a package with this file included:
https://github.com/openwrt/openwrt/blob/0dd5505a17959c0b3eabac2827031487444708fb/package/kernel/linux/modules/netfilter.mk#L1193
This would mean that we could not have both packages installed.

Will need to break the duplicated Kconfigs and files out into another package, and depend on that from both `nft-dup` and `nft-netdev`.
You will also need to address the `CONNTRACK` dependency: https://elixir.bootlin.com/linux/v5.15.111/source/net/ipv4/netfilter/Kconfig#L38

---
Probably best to wait for a second opinion before starting these next parts:

Could also define a Make variable for the dup Kconfigs, and files: https://github.com/openwrt/openwrt/blob/0dd5505a17959c0b3eabac2827031487444708fb/include/netfilter.mk, but given the mix in the package Makefile, it seems like that may no longer be needed.

`ipt-tee` does include the `NF_DUP_IPV*` symbols (being an iptables package, not worried about that conflict), but you could break those out into their own Make variables (in the include file), and packages.
https://github.com/openwrt/openwrt/blob/0dd5505a17959c0b3eabac2827031487444708fb/package/kernel/linux/modules/netfilter.mk#L717
https://github.com/openwrt/openwrt/blob/0dd5505a17959c0b3eabac2827031487444708fb/include/netfilter.mk#L257

kwiateusz replied:
Thx for comment. I didn't know that this will make conflict so as you suggested I've splitted this into separate packages.

About second part have you any docs/links that suggests how to do that? Because as I'm looking into that file I can't figure out how variables works there and I'm not very familiar with makefile syntax,

john-tho replied:
About using the netfilter include variables:
No _need_ to use them (unless a committer wants you to).
To learn about it:
`nf_add` is a Makefile function / macro which appends to variables (name given by a parameter) when called: `$(call nf_add, variable-name, Kconfig, module-filename-no-extension, optional-kernel-version-condition)` gives `KCONFIG_variable-name`, and `variable-name-$(Kconfig-state)`, so you get `-m` suffix for modules: `variable-name-m` (a list).
These variables are then refered to in the netfilter package Makefile to get the paths of `.ko` files, and the modprobe names using the `-m` suffix variable, so that the packages do not miss files if the Kconfig is built-in to kernel (that would create a `-y` suffix variable).
Good example usage in a previous commit: https://github.com/openwrt/openwrt/commit/144fa4d4e7c00df370924751f58781689cd333e7

ptpt52 commented:
what is the use case for this dup kmod

champtar commented:
@kwiateusz have you tried just installing `kmod-nft-netdev` ?

kwiateusz commented:
@champtar yes I've tried and can't use rules like
`pkttype broadcast dup to 255.255.255.255 device br-lan.76 notrack counter
   
ip daddr 224.0.0.251 udp dport mdns notrack counter dup to 224.0.0.251 device br-lan.76 comment "Forwad mDNS traffic"`

because of missing modules

@ptpt52 as I showed above you can for example push broadcast form one vlan to another only using netfilter

champtar commented:
> @champtar yes I've tried and can't use rules like
> 
> pkttype broadcast dup to 255.255.255.255 device br-lan.76 notrack counter
> 
> ip daddr 224.0.0.251 udp dport mdns notrack counter dup to 224.0.0.251 device br-lan.76 comment "Forwad mDNS traffic"

Can you try `dup to br-lan.76` ?

> 
> because of missing modules
> 
> Here is thread with that same problem https://forum.openwrt.org/t/firewall4-custom-nftables-rule-with-dup-statement-results-in-unknown-family/139043/2
> 
> @ptpt52 as I showed above you can for example push broadcast form one vlan to another using only netfilter

kwiateusz commented:
@champtar in phantapp is used netdev table and that can work with kmod-netdev
 but I've used `table ip` and there dup_ip modules are needed

So this change allow to duplicate packets in different table types.

champtar commented:
> @champtar in phantapp is used netdev table and that can work with kmod-netdev but I've used `table ip` and there dup_ip modules are needed
> 
> So this change allow to duplicate packets in different table types.

Can you update the commit message to explain that dup was already available for netdev family but not for the ip family, I started commenting only looking at the commit message and this was not clear

kwiateusz committed:
netfilter: add kmod-nft-dup-inet

Add kmod-nft-dup-inet package to allow packet duplication in ip/ip6/inet nftables family

Signed-off-by: Michał Kwiatek <michal@kwiatek.it>

kwiateusz commented:
@champtar thanks for your messages, I've changed whole kmod to include only needed packets for ip and inet familly as netdev has own package. And I hope that commit message is now appropriate

john-tho commented:
Still need the `conntrack` dependency for the ipv4 and ipv6 modules:
https://elixir.bootlin.com/linux/v5.15.111/source/net/ipv4/netfilter/nf_dup_ipv4.c#L20
https://elixir.bootlin.com/linux/v5.15.111/source/net/ipv6/netfilter/nf_dup_ipv6.c#L16
This is because the buildbots enable all kernel modules, so the kernel is built with `CONFIG_NF_CONNTRACK=m` (from `kmod-nf-conntrack`), so `dup_ipv*` will depend on it: `depends on !NF_CONNTRACK || NF_CONNTRACK`, but you will not see this dependency if you do not build `NF_CONNTRACK`.

kwiateusz commented:
nft-core depends on conntrack isn't that enough? I build that locally and everything went fine on ipq807x

kwiateusz commented:
@john-tho I've just run this mod on ipq807x (Xiaomi ax3600) and everything works as expected

john-tho commented:
Note that I'm not a _committer_, so my opinion does not matter much, nor could I get this merged.

> nft-core depends on conntrack isn't that enough?

Sort of. From `nft-core`, `conntrack6` dependency is conditional on the `@IPV6` symbol being set. While OpenWrt now always enables this: https://github.com/openwrt/openwrt/commit/832e7b817221d288df76b763ca12c585365db5d8, there are still all the conditionals for it.
I would depend on `kmod-nf-conntrack` so that it is clear.
Maybe also depend on `@IPV6`.
Best to wait for a committer to give some guidance.

kwiateusz commented:
@Ansuel sorry to bother you but I see you're commiting a lot of cool stuff. So maybe could you give me a hint if I need to do something else, ping someone or just wait for luck that someone with the right permissions will happen to look here?

kwiateusz commented:
Thx for merge in https://github.com/openwrt/openwrt/commit/a7e9445975f832db887e6044d7e84220d2a68cf1
